url026791d51b_dhleong_SeriesTracker_v6_tgz-pJ8-stv6_SeriesTrackerJ8/src/stv6/http/HttpServer.java

```diff
@@ -0,0 +1,157 @@
+package stv6.http;
+
+import java.io.IOException;
+import java.net.ServerSocket;
+import java.net.Socket;
+import java.text.ParseException;
+import java.text.SimpleDateFormat;
+import java.util.Date;
+import java.util.TimeZone;
+
+import stv6.http.pools.ClientPool;
+import stv6.http.request.RequestHandler;
+import stv6.http.request.RequestHandlerManager;
+import stv6.http.request.Response;
+
+public abstract class HttpServer implements Runnable {
+
+	private static SimpleDateFormat dateFormatter 
+		= getDateFormatter();
+	
+	static {
+		// make sure it's in GMT
+		dateFormatter.setTimeZone(TimeZone.getTimeZone("GMT"));
+	}
+	
+	private boolean running;
+	private Thread thisThread;
+	
+	private ServerSocket serverskt;
+	private final ClientPool clients;
+	private final RequestHandlerManager responder;
+		
+	public HttpServer(int port) throws IOException {
+		running = false;
+		
+		setPort(port);
+		clients = new ClientPool();
+		responder = getHandlerManager();
+	}
+	
+	/**
+	 * Implement this to use your own client
+	 * 
+	 * @param socket
+	 * @param callback
+	 * @return
+	 */
+	public abstract Client createClient(HttpSocket socket, RequestHandlerManager callback);
+	
+	/**
+	 * Overwrite this if you want to extend RHM somehow
+	 */
+	public RequestHandlerManager getHandlerManager() {
+		return new RequestHandlerManager();
+	}
+
+	public static void generate404(Response resp) {		
+		resp.setStatus("404 Not Found");
+		resp.setContentType("text/html");
+		resp.append("<!DOCTYPE HTML PUBLIC \"-//W3C//DTD HTML 4.01//EN\""
+				+" \"http://www.w3.org/TR/html4/strict.dtd\">\n");
+		resp.append("<html>\n<head>\n<title>404 File not found</title>\n");
+		resp.append("<meta http-equiv=\"Content-Type\" content=\"text/html;charset=utf-8\" >\n");
+		resp.append("<link rel=\"stylesheet\" href=\"files/css/default.css\" type=\"text/css\">\n");
+		resp.append("</head>\n<body>\n");
+		resp.append("<h1>404 File not found</h1>I could not find the page you requested.");
+		resp.append(" Please go back and try again.");
+		resp.append("\n</body>\n</html>");
+	}
+	
+	public int getPort() {
+		if (serverskt != null)
+			return serverskt.getLocalPort();
+		
+		return -1;
+	}
+	
+	private Socket getSocket() {
+		try {
+			return serverskt.accept();
+		} catch (IOException e) {
+			return null;
+		}
+	}
+	
+	public void registerHandler(RequestHandler handler) {
+		responder.registerHandler(handler);
+	}
+	
+	@Override
+	public void run() {
+		while (thisThread == Thread.currentThread() && running) {
+			Socket sock = getSocket();
+			if (sock == null) {
+				try {
+					Thread.sleep(10);
+				} catch (InterruptedException e) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: InterruptedException is ignored. This can delay thread shutdown and clear the thread’s interrupt status. Only code that implements a thread’s interruption policy can swallow an interruption request.

**Fix**: Rethrow the InterruptedException or reinterrupt the current thread using *Thread.currentThread().interrupt()* so that higher-level interrupt handlers can function correctly.
 If you are wrapping the InterruptedException inside a RuntimeException, call *Thread.currentThread().interrupt()* before throwing the RuntimeException.

Learn more about [interrupts](https://docs.oracle.com/javase/tutorial/essential/concurrency/interrupt.html) and [dealing with InterruptedException](https://www.ibm.com/developerworks/java/library/j-jtp05236/index.html?ca=drs-#N10187)